crypto/src/asset/amount.rs

```diff
@@ -117,85 +211,110 @@ impl From<Amount> for Fq {
     }
 }
 
-impl From<u64> for Amount {
-    fn from(amount: u64) -> Amount {
+impl From<Amount> for Fr {
+    fn from(amount: Amount) -> Fr {
+        Fr::from(amount.inner)
+    }
+}
+
+impl From<u128> for Amount {
+    fn from(amount: u128) -> Amount {
         Amount { inner: amount }
     }
 }
 
-impl From<u32> for Amount {
-    fn from(amount: u32) -> Amount {
+impl From<Amount> for u128 {
+    fn from(amount: Amount) -> u128 {
+        amount.inner as u128
+    }
+}
+
+impl From<i128> for Amount {
+    fn from(amount: i128) -> Amount {
         Amount {
-            inner: amount as u64,
+            inner: amount as u128,
         }
     }
 }
 
-impl From<Amount> for u32 {
-    fn from(amount: Amount) -> u32 {
-        amount.inner as u32
+impl From<Amount> for i128 {
+    fn from(amount: Amount) -> i128 {
+        amount.inner as i128
     }
 }
 
-impl From<i64> for Amount {
-    fn from(amount: i64) -> Amount {
+impl From<[u8; 16]> for Amount {
+    fn from(bytes: [u8; 16]) -> Amount {
         Amount {
-            inner: amount as u64,
+            inner: u128::from_le_bytes(bytes),
         }
     }
 }
 
-impl From<Amount> for u128 {
-    fn from(amount: Amount) -> u128 {
-        amount.inner as u128
+impl TryFrom<Vec<u8>> for Amount {
+    type Error = anyhow::Error;
+
+    fn try_from(bytes: Vec<u8>) -> Result<Self, Self::Error> {
+        if bytes.len() != 16 {
+            return Err(anyhow::anyhow!(
+                "could not deserialize Amount: input vec is not 16 bytes"
+            ));
+        }
+        let mut array = [0u8; 16];
+        array.copy_from_slice(&bytes);
+        Ok(Amount {
+            inner: u128::from_le_bytes(array),
+        })
+    }
+}
```

Comment:
Ahhh, the proto format is variable-length, but we need a fixed-length encoding for use in ciphertexts.

Reply:
Maybe, in order to avoid confusion, rather than having a `From/TryFrom` pair here, we could have ordinary methods on the `Amount` type like

```
Amount::to_fixed_len_bytes(&self) -> [u8; 16]
Amount::from_fixed_len_bytes(&[u8]) -> Result<Self, _>
```
so that it's clear what the distinction is from ordinary serialization, and we can write a comment on the methods.

Reply:
+1, the `Vec<u8>` conversions weren't actually used anywhere, so I removed them and use `to_le_bytes`/`from_le_bytes` as passthrough to `u128` instead to take care of the `Note` usecase